File: daq/slc/dqm/src/Histo2.cc
#include "daq/slc/dqm/Histo2.h"

using namespace Belle2;

Histo2::Histo2() throw() : Histo() {}

Histo2::Histo2(const Histo2& h) throw() : Histo(h) {}

Histo2::Histo2(const std::string& name, const std::string& title,
               int nbinx, double xmin, double xmax,
               int nbiny, double ymin, double ymax) throw() : Histo(name, title)
{
  m_axis_x.set(nbinx, xmin, xmax);
  m_axis_y.set(nbiny, ymin, ymax);
}

void Histo2::reset() throw()
{
  setUpdateId(0);
  m_data->setAll(0);
}

void Histo2::resetAxis(int nbinx, double xmin, double xmax,
                       int nbiny, double ymin, double ymax)
{
  m_axis_x.set(nbinx, xmin, xmax);
  m_axis_y.set(nbiny, ymin, ymax);
  m_data->resize((m_axis_x.getNbins() + 2) * (m_axis_y.getNbins() + 2));
}

double Histo2::getBinContent(int nx, int ny) const throw()
{
  return m_data->get((nx + 1) + (ny + 1) * (m_axis_x.getNbins() + 2));
}

double Histo2::getOverFlow() const throw() { return m_data->get(m_data->size() - 1); }

double Histo2::getUnderFlow() const throw() { return m_data->get(0); }

double Histo2::getOverFlowX(int) const throw() { return m_data->get(m_data->size() - 1); }

double Histo2::getUnderFlowX(int) const throw() { return m_data->get(0); }

double Histo2::getOverFlowY(int) const throw() { return m_data->get(m_data->size() - 1); }

double Histo2::getUnderFlowY(int) const throw() { return m_data->get(0); }

void Histo2::setBinContent(int nx, int ny, double data) throw()
{
  m_data->set((nx + 1) + (ny + 1) * (m_axis_x.getNbins() + 2), data);
}

void Histo2::setOverFlow(double data) throw() { m_data->set(m_data->size() - 1, data); }

void Histo2::setOverFlowX(int, double data) throw() { m_data->set(m_data->size() - 1, data); }

void Histo2::setOverFlowY(int, double data) throw() { m_data->set(m_data->size() - 1, data); }

void Histo2::setUnderFlow(double data) throw() { m_data->set(0, data); }

void Histo2::setUnderFlowX(int, double data) throw() { m_data->set(0, data); }

void Histo2::setUnderFlowY(int, double data) throw() { m_data->set(0, data); }

double Histo2::getMaximum() const throw() { return m_axis_z.getMax(); }

double Histo2::getMinimum() const throw() { return m_axis_z.getMin(); }

void Histo2::setMaximum(double data) throw()
{
  if (!m_axis_z.isFixedMax()) m_axis_z.setMax(data);
}

void Histo2::setMinimum(double data) throw()
{
  if (!m_axis_z.isFixedMin()) m_axis_z.setMin(data);
}

void Histo2::fixMaximum(double data) throw()
{
  m_axis_z.fixMax(true);
  m_axis_z.setMax(data);
}

void Histo2::fixMinimum(double data) throw()
{
  m_axis_z.fixMin(true);
  m_axis_z.setMin(data);
}

void Histo2::readConfig(Reader& reader) throw(IOException)
{
  Histo::readConfig(reader);
  m_axis_x.readObject(reader);
  m_axis_y.readObject(reader);
  m_data->resize((m_axis_x.getNbins() + 2) * (m_axis_y.getNbins() + 2));
}

void Histo2::writeConfig(Writer& writer) const throw(IOException)
{
  Histo::writeConfig(writer);
  m_axis_x.writeObject(writer);
  m_axis_y.writeObject(writer);
}

File: daq/slc/extra/javalib/src/b2daq/dqm/core/Histo2.java
package b2daq.dqm.core;

import java.io.IOException;

import b2daq.core.Reader;
import b2daq.core.Writer;

public abstract class Histo2 extends Histo {

	private boolean _stat = true;
	
	public Histo2() {
		super();
	}

	public Histo2(String name, String title, int nbinx, double xmin, double xmax
			, int nbiny, double ymin, double ymax) {
		super(name, title);
		_axis_x.setRange(nbinx, xmin, xmax);
		_axis_y.setRange(nbiny, ymin, ymax);
	}
/*
	public void reset() {
		super.reset();
		setMaximum(1);
	}
*/	
	public boolean getStat() { return _stat; }
	public void setStat(boolean stat) { _stat = stat; }
	
	public void fixMaximum(double data, boolean fix) {
		_axis_z.setMax(data);
		_axis_z.fixMax(fix);
	}

	public void fixMinimum(double data, boolean fix) {
		_axis_z.setMin(data);
		_axis_z.fixMin(fix);
	}
	public double getBinContent(int nx, int ny) { 
		return _data_v.get((nx+1) + (ny+1)*(_axis_x.getNbins()+2));
	}
	public double getOverFlow() { return _data_v.get(_data_v.length() - 1); }
	public double getUnderFlow() { return _data_v.get(0); }
	public double getOverFlowX(int ny) { return _data_v.get(_data_v.length() - 1); }
	public double getUnderFlowX(int ny) { return _data_v.get(0); }
	public double getOverFlowY(int nx) { return _data_v.get(_data_v.length() - 1); }
	public double getUnderFlowY(int nx) { return _data_v.get(0); }
	public void setBinContent(int nx, int ny, double data) {
		if ( nx >= 0 && nx < _axis_x.getNbins() && ny >= 0 && ny < _axis_y.getNbins()) {
			_data_v.set((nx+1) + (ny+1)*(_axis_x.getNbins()+2), data);
			if ( data * 1.05 > getMaximum() ) setMaximum(data * 1.05);
		}
	}
	public void setOverFlow(double data) { _data_v.set(_data_v.length() - 1, data); }
	public void setUnderFlow(double data) { _data_v.set(0, data); }
    
	public int getDim() { return 2; }
	public double getMaximum() { return _axis_z.getMax(); }
	public double getMinimum() { return _axis_z.getMin(); }
	public void setMaximum(double data) {
		if ( !_axis_z.isFixMax() ) _axis_z.setMax(data);
	}
	public void setMinimum(double data) {
		if ( !_axis_z.isFixMin() ) _axis_z.setMin(data);
	}

	public void setMaxAndMin(){
		if ( ! _axis_y.isFixMax() || ! _axis_y.isFixMin()) {
			double data; 
			for ( int ny = 0;  ny < _axis_y.getNbins(); ny++) {
				for ( int nx = 0;  nx < _axis_x.getNbins(); nx++) {
					data = getBinContent(nx, ny);
					if ( data * 1.05 > getMaximum() ) setMaximum(data * 1.05);
				}
			}
		}
	}

	public double getMeanX() {
		int entries = 0;
		double bin = 0;
		double mean = 0;
		for (int nx = 0; nx < _axis_x.getNbins(); nx++) {
			bin = 0;
			for (int ny = 0; ny < _axis_y.getNbins(); ny++) {
				bin += getBinContent(nx, ny);
			}
			mean += bin * ((nx + 0.5) * (_axis_x.getMax() - _axis_x.getMin())
					/ _axis_x.getNbins() + _axis_x.getMin());
			entries += bin;
		}
		return mean / entries;
	}
	
	public double getMeanY() {
		int entries = 0;
		double bin = 0;
		double mean = 0;
		for (int ny = 0; ny < _axis_y.getNbins(); ny++) {
			bin = 0;
			for (int nx = 0; nx < _axis_x.getNbins(); nx++) {
				bin += getBinContent(nx, ny);
			}
			mean += bin * ((ny + 0.5) * (_axis_y.getMax() - _axis_y.getMin())
					/ _axis_y.getNbins() + _axis_y.getMin());
			entries += bin;
		}
		return mean / entries;
	}
	
	public double getEntries() {
		int entries = 0;
		for ( int ny = 0;  ny < _axis_y.getNbins(); ny++) {
			for ( int nx = 0;  nx < _axis_x.getNbins(); nx++) {
				entries += getBinContent(nx, ny);
			}
		}
		return entries;
	}

	@Override
	public void readConfig(Reader reader) throws IOException {
		setTabId(reader.readChar());
		setPositionId(reader.readChar());
		setName(reader.readString());
		setTitle(reader.readString());
		setAxisX(readAxis(reader));
		setAxisY(readAxis(reader));
		_data_v.resize((getAxisX().getNbins()+2)*(getAxisY().getNbins()+2));
	}

	@Override
	public void writeConfig(Writer writer) throws IOException {
		writer.writeString(getDataType());
		writer.writeChar((char)getTabId());
		writer.writeChar((char)getPositionId());
		writer.writeString(getName());
		writer.writeString(getTitle());
		writeAxis(writer, getAxisX());
		writeAxis(writer, getAxisY());
	}

	protected String getXML() {
		String str = "title=\"" + getTitle() + 
		"\" data=\"" + _data_v.toString() + 
		"\" >\n";
		str += _axis_x.getXML("axis-x");
		str += _axis_y.getXML("axis-y");
		str += _axis_z.getXML("axis-z");
		return str;
	}

	public Histo clone() {
		try {
			Histo2 h = (Histo2) HistoFactory.create(getDataType());
			h.setName(getName());
			h.setTitle(getTitle());
			h.getAxisX().copy(getAxisX());
			h.getAxisY().copy(getAxisY());
			h.getAxisZ().copy(getAxisZ());
			h.getData().copy(getData());
			return h;
		} catch (WrongDataTypeException e) {
			e.printStackTrace();
			return null;
		} 
	}

	public void add(Histo h, double scale) {
		if (h.getAxisY().getNbins() == getAxisY().getNbins() &&
				h.getAxisY().getMin() == getAxisY().getMin() && 
				h.getAxisY().getMax() == getAxisY().getMax() &&
				h.getAxisX().getNbins() == getAxisX().getNbins() &&
				h.getAxisX().getMin() == getAxisX().getMin() && 
				h.getAxisX().getMax() == getAxisX().getMax() ) {
			for (int ny = 0; ny < getAxisY().getNbins(); ny++) {
				for (int nx = 0; nx < getAxisX().getNbins(); nx++) {
					setBinContent(nx, ny, getBinContent(nx, ny) + h.getBinContent(nx, ny) * scale);
				}
			}
		}
	}

}
